### PR TITLE
Add CUDA graph support for torch.while_loop

### PR DESCRIPTION
**Pull request by galv**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #186056
* __->__ #186055

Motivation is to allow us to handle some forms of dynamic shapes
inside of a cuda graph. This does not necessarily improve performance,
but it can make your entire workload cuda graph compatible, which is
good for developer experience

For example, if you want to do a reduction across an embedding table,
you can do that with a variable length input index tensor. This will
work on the basis that the output shape is always the same.

Another motivating example is applying a loss to the variable number
of sequences packed into a larger sequence, which currently requires
copying data from device to host in this implementation:
https://github.com/NVIDIA-NeMo/RL/blob/ab079b81f80fa10754c29ac99ec91263c77ec8fa/nemo_rl/algorithms/loss/wrapper.py#L87-L88
There, the loss scalar output is always the same.

Finally, there is potential to do MoE using a while loop, since it has
the property that it has dynamic shapes in the middle of its
computation, but fixed shapes at the end. However, this is quite
complicated for training due to (1) torch.while_loop()'s need to have
a fixed maximum trip count and (2) the increase in memory usage
implied by cached activations for every iteration of the unrolled
while loop. Finally, it is unlikely to be performant anyway.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Also, would now capturing RNN or optimization/sampling loops should be possible if loops can be captured in CUDA graphs using efficient inplace?

**Comment by zou3519** [comment]
@claude review these changes

**Comment by zou3519** [comment]
@claude review these changes

**Review by eellison — APPROVED** [review]
minimal changes from previously accepted pr - https://github.com/pytorch/pytorch/pull/140979 , accepting

**Comment by eellison** [comment]
maybe rebase to more green main though..

**Comment by galv** [comment]
@eellison yep, I agree. I should have some time today or tomorrow to rebase this and clean it up.

**Comment by galv** [comment]
@pytorchbot merge -i

The one failure is a geenric infra error.

**Comment by galv** [comment]
@pytorchbot merge -i

The one failure is a generic infra error.
